### PR TITLE
Lock composer requires to PATCH versions only

### PR DESCRIPTION
**Pull request by firstred**
| Questions | Answers |
| --- | --- |
| Branch? | `develop` |
| Description? | I noticed that the libraries in `composer.json` are sometimes allowed to receive MINOR or even MAJOR upgrades. I'm not sure if that was intentional and I can imagine that you'd rather keep updates limited to PATCH only in order to keep everything compatible. |
| Type? | improvement |
| Category? | CO |
| BC breaks? | no |
| Deprecations? | no |
| Fixed ticket? |  |
| How to test? |  |

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -10,29 +10,29 @@
...
-    "shudrum/array-finder": "1.1.0",
-    "phpoffice/phpexcel": "1.8",
-    "guzzlehttp/guzzle": "~4.0",
-    "csa/guzzle-bundle": "~1.3",
-    "paragonie/random_compat": "^1.4|^2.0",
+    "shudrum/array-finder": "~1.1.0",
+    "phpoffice/phpexcel": "~1.8.0",
+    "guzzlehttp/guzzle": "~4.0.0",
+    "csa/guzzle-bundle": "~1.3.0",
+    "paragonie/random_compat": "~2.0.0",
```
Comment (mickaelandrieu) [inline composer.json]: Keep the 1.4 because the version 2 requires PHP 7+ afaik

File: composer.json [hunk 2]
```diff
@@ -91,11 +91,11 @@
...
   },
   "require-dev": {
-    "phpunit/phpunit": "~4.5",
+    "phpunit/phpunit": "~4.5.0",
     "symfony/phpunit-bridge": "~3.1.3",
     "phake/phake": "@stable",
-    "friendsofphp/php-cs-fixer": "^1.10",
-    "sensio/generator-bundle": "~2.3"
+    "friendsofphp/php-cs-fixer": "~1.10.0",
+    "sensio/generator-bundle": "~2.3.0"
```
Comment (mickaelandrieu) [inline composer.json]: You can drop generator bundle which is not used anymore.

File: composer.json (hunk 1 above)
Comment (firstred) [inline composer.json]: Ahh, it's good that we have had a look at it, then.

File: composer.json [hunk 3]
```diff
@@ -10,29 +10,29 @@
...
-    "shudrum/array-finder": "1.1.0",
-    "phpoffice/phpexcel": "1.8",
-    "guzzlehttp/guzzle": "~4.0",
-    "csa/guzzle-bundle": "~1.3",
-    "paragonie/random_compat": "^1.4|^2.0",
+    "shudrum/array-finder": "~1.1.0",
+    "phpoffice/phpexcel": "~1.8.0",
+    "guzzlehttp/guzzle": "~4.0.0",
+    "csa/guzzle-bundle": "~1.3.0",
+    "paragonie/random_compat": "~1.4.0",
```
Comment (mickaelandrieu) [inline composer.json]: @firstred  "~1.4.0|~2.0.0"

Comment (firstred) [inline composer.json]: Do you know what the pipe characters means here?

I always thought it was an `OR`, but I couldn't really figure it out. We don't want any 2.0 stuff, do we?

Comment (firstred) [inline composer.json]: I can't find the meaning of it at all :confused: 
https://getcomposer.org/doc/articles/versions.md

Comment (firstred) [inline composer.json]: This page explains that it is the same as double-pipe `||` or `OR`: http://qpleple.com/understand-composer-versions/

But why do that if we can't go for the `PHP 7.0` only version (`2.0`)?

Comment (mickaelandrieu) [inline composer.json]: I'm in doubt, let's ask to package maintainer what to do => https://github.com/paragonie/random_compat/issues/116

Comment (mickaelandrieu) [inline composer.json]: So, for now only keep 1.4.0 please, thank you :)

File: composer.json (hunk 1 above)
Comment (Uebix) [inline composer.json]: http://forge.prestashop.com/browse/BOOM-1773

File: composer.json (hunk 3 above)
Comment (Uebix) [inline composer.json]: Yes, please use only paragonie/random_compact ^1.4
http://forge.prestashop.com/browse/BOOM-1773

Comment (firstred) [inline composer.json]: :open_mouth: And this was not looked at before because the branch had conflicts?

**Comment by firstred** [comment]
Does anyone know why there are two minifier libraries?
```
    "mrclay/minify": "~2.3.0",
    "matthiasmullie/minify": "~1.3.0",
```
I also don't understand the `OR`
How does composer determine the right version, when it is unknown which PHP version is going to be used when preparing the PrestaShop package?  
Example:
```
"doctrine/cache": "~1.6|~1.5",
```

**Comment by Quetzacoalt91** [comment]
Hello @firstred,

> How does composer determine the right version, when it is unknown which PHP version is going to be used when preparing the PrestaShop package?

The issue happened when I added the `doctrine/cache` into the project and tested it on Travis.
The version 1.6 of this component works on PHP 5.5, 5.6 and 7, but not 5.4. That's why I added the fallback. :)
Anyway, when we reease a new version, we now use a 5.4 binary of PHP.

File: composer.json [hunk 4]
```diff
@@ -14,30 +14,30 @@
...
-    "shudrum/array-finder": "1.1.0",
-    "phpoffice/phpexcel": "1.8",
-    "guzzlehttp/guzzle": "~5.0",
-    "csa/guzzle-bundle": "~1.3",
-    "paragonie/random_compat": "^1.4|^2.0",
+    "shudrum/array-finder": "~1.1.0",
+    "phpoffice/phpexcel": "~1.8.0",
+    "guzzlehttp/guzzle": "~5.0.0",
+    "csa/guzzle-bundle": "~1.3.0",
+    "paragonie/random_compat": "~1.4.0",
```
Comment (Quetzacoalt91) [inline composer.json]: Hello,

I'm currently testing this PR which can solve an issue.
Unfortunately, this change break some relations:
```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - defuse/php-encryption 2.0.3 requires paragonie/random_compat ~2.0 -> satisfiable by paragonie/random_compat[v2.0.0, v2.0.1, v2.0.2, v2.0.3, v2.0.4] but these conflict with your requirements or minimum-stability.
    - defuse/php-encryption 2.0.1 requires paragonie/random_compat ~2.0 -> satisfiable by paragonie/random_compat[v2.0.0, v2.0.1, v2.0.2, v2.0.3, v2.0.4] but these conflict with your requirements or minimum-stability.
    - defuse/php-encryption 2.0.3 requires paragonie/random_compat ~2.0 -> satisfiable by paragonie/random_compat[v2.0.0, v2.0.1, v2.0.2, v2.0.3, v2.0.4] but these conflict with your requirements or minimum-stability.
    - Installation request for defuse/php-encryption ~2.0.1 -> satisfiable by defuse/php-encryption[2.0.1, 2.0.3].
```

**Pull request closed without merging**